src/webgpu/shader/execution/builtin/builtin.ts

```diff
@@ -89,169 +80,389 @@ export function submitComputeShader(
   pass.endPass();
 
   t.queue.submit([encoder.finish()]);
-  return outputBuffer;
-}
-
-export function createBuiltinCall(
-  baseType: ScalarType,
-  arrayLength: number,
-  builtin: string,
-  outputDataAsF32: Float32Array,
-  inputData: TypedArrayBufferView,
-  cases: Cases,
-  numberType: OperandType
-): string[] {
-  let inputDataAsF32: Float32Array;
-  let inputDataAsU32: Uint32Array;
-  let inputDataAsI32: Int32Array;
-  if (numberType === OperandType.Float) {
-    inputDataAsF32 = inputData as Float32Array;
-    inputDataAsU32 = new Uint32Array(inputData.buffer);
-    inputDataAsI32 = new Int32Array(inputData.buffer);
-  } else {
-    inputDataAsF32 = new Float32Array(inputData.buffer);
-    inputDataAsU32 = inputData as Uint32Array;
-    inputDataAsI32 = new Int32Array(inputData.buffer);
-  }
 
-  const outputDataAsU32 = new Uint32Array(outputDataAsF32.buffer);
-  const outputDataAsI32 = new Int32Array(outputDataAsF32.buffer);
-
-  const errs: string[] = [];
-  for (let i = 0; i < cases.length; i++) {
-    const input: string[] = [];
-    const output: string[] = [];
-    const expected: string[] = [];
-    let matched = true;
-    for (let j = 0; j < arrayLength; j++) {
-      const idx = i * 4 + j;
-      const caseExpected: string[] = [];
-      for (const e of cases[i].expected) {
-        const expectedDataAsU32 = numberType === OperandType.Float ? float32ToUint32(e) : e;
-        const expectedDataAsF32 = numberType === OperandType.Float ? e : uint32ToFloat32(e);
-        const expectedDataAsI32 =
-          numberType === OperandType.Float ? float32ToInt32(e) : uint32ToInt32(e);
-
-        switch (baseType) {
-          case 'u32': {
-            caseExpected.push(expectedDataAsU32 + ' (0x' + expectedDataAsU32.toString(16) + ')');
-            if (outputDataAsU32[idx] !== expectedDataAsU32) {
-              matched = false;
-            }
-            break;
-          }
-          case 'i32': {
-            caseExpected.push(expectedDataAsI32 + ' (0x' + expectedDataAsU32.toString(16) + ')');
-            if (outputDataAsI32[idx] !== expectedDataAsI32) {
-              matched = false;
-            }
-            break;
+  const checkExpectation = (outputData: typeof inputData) => {
```

Comment:
optional: maybe it's possible now to use the existing `checkElementsEqual`? (not sure)

Reply:
It is possible, the difference is that checkExpectation outputs better debug info in case of failure. I'll change it if you still think it's best to use checkElemetsEqual.

Reply:
Ah, in that case no need, unless you think checkElementsEqual can be improved to output the better debug info.